File: problem3.rb
#!/usr/bin/env ruby

n = ARGV[0].to_i
exit unless n

def factorize(n, factors = [])
	max_factor = Math.sqrt(n).floor
	(2..max_factor).each do |i|
		if n % i == 0
			factors.push(i)
			n /= i
			return factorize(n, factors)
		end
	end

	factors.push(n)
end

factors = factorize(n)

puts "The largest prime factor of the number #{n} is #{factors[-1]}"

File: problem3.py
#!/usr/bin/env python

import math
def factorize(n, factors = None):
	if factors is None:
		factors = []
	
	max_factor = int(math.floor(math.sqrt(n))) + 1
	for i in xrange(2, max_factor):
		if n % i == 0:
			factors.append(i)
			n /= i
			return factorize(n, factors)
	factors.append(n)
	return factors

if __name__ == "__main__":
	import sys
	n = int(sys.argv[1])
	if not n:
		exit(1)


	factors = factorize(n)
	print 'The largest prime factor of the number', n, 'is', factors[-1]
